### ISSUE TITLE
Port 16-bit audio support via AHI from Minimig

### ISSUE THREAD

**Issue by limi** [MiSTer-devel/Minimig-AGA_MiSTer#171]
[Now that we have RTG support available](https://misterfpga.org/viewtopic.php?p=12078#p12078) in MiSTer, the next missing piece of the classic late-stage Amiga setup would be… sound card support. An Amiga 4000 with 24-bit graphics card (RTG) and 16-bit sound card (AHI) was a reasonably common setup late in the Amiga’s lifecycle.

For those of you who don’t know, Amiga had support for 16-bit audio interfaces like Delfina, Prelude, Toccata, etc. It was the “retargetable audio” interface for the Amiga, where the hardware was abstracted away, and allowed AmigaOS to address a set of Amiga sound cards. It was later included in official AmigaOS releases.

Now, [Paul Honig has implemented the 16-bit Toccata Amiga sound card in FPGA](https://misterfpga.org/viewtopic.php?p=82771#p82771) (SystemVerilog), and hooked it up to Minimig. (Note: this is playback only, sound input is way less important)

GitHub repository is here:
https://github.com/ranzbak/fpga-toccata

And an implementation example:
https://github.com/ranzbak/MinimigAGA_TC64/tree/v5.0/rtl/openaars/toccata

Any chance we could get this supported in the MiSTer core?

**Comment by sorgelig** [comment]
What is the usage scenario of it?
Games don't know about this card i think. AmigaOS system sounds would be just some short samples like ding, dong.. Will it be beneficial from 16bit?

**Comment by limi** [comment]
The main use case would be to play music with higher quality, yes. Since the Amiga is very heavy on demo scene productions (still!), some of them want RTG and AHI to work. (There are a few games, but I agree that’s not the main use case)

OctaMED (and other smaller music apps) supports it, the main music players have AHI mixers that will improve the quality if the samples are 12/14/16-bit, etc. It also lets those players output e.g. S3M modules without downgrading the quality. There was a fair bit of demo scene crossover with the PC in that part of Amiga’s history, and we used to play S3M music often.

**Comment by limi** [comment]
Oh, and I don’t know if the better approach for MiSTer specifically is to hand it over to a different part of the system, like when RTG was implemented using the framebuffer instead of implementing an old RTG graphics card.

Not sure if you can output 16-bit audio directly from the sound driver and skip the FPGA implementation of the sound chip itself (keeping the Toccata outputs so you can use its AHI driver) and use the one on the Linux side — but that might be better, if possible.

**Comment by limi** [comment]
Looks like @robinsonb5 has added it to MiST:

https://www.atari-forum.com/viewtopic.php?p=458089#p458089

**Comment by robinsonb5** [comment]
@ranzbak already implemented it in his own fork of the MiST Minimig core (targetted at a Xilinx-based board) - all I needed to do was backport his changes - everything required is in this commit:
https://github.com/minimig-dev/MinimigAGA-MiST-TC64/commit/c63f8cb6763ecbc4e9fa50ef89b698d3d5106e98
though I think the autoconfig stuff will need adapting for MiSTer.
If no-one's already working on it I can take a look.

**Comment by sorgelig** [comment]
> If no-one's already working on it I can take a look.

Please go on. This feature needs a testbench in AmigaOS with all drivers/apps installed which i don't have yet.

**Comment by robinsonb5** [comment]
Here's an RBF for testing - source here: https://github.com/robinsonb5-PRs/Minimig-AGA_MiSTer/tree/toccata

The fpga-toccata submodule is currently pulled from my robinsonb5-PRs organisation, incorporating retrofun's fix for mono playback and a provisional fix for a FIFO issue which I encountered.

Once Paul's happy with the FIFO fix (or if he comes up with a better way to solve the problem) then I'll point the submodule back to his upstream repo.

[Minimig_toccata_20240215.zip](https://github.com/MiSTer-devel/Minimig-AGA_MiSTer/files/14303366/Minimig_toccata_20240215.zip)

**Comment by apolkosnik** [comment]
Works for me. I've installed the v9.10 drivers, AHI, the  v12 library, and also applied a toclib12.lha patch from Aminet. It seems to work with EaglePlayer, and the modes show up in AHI.

**Comment by chris-y** [comment]
I've installed toclib12 and the driver from m68k-amigaos-ahidrv - test sound works in AHI.
I've run OctaMED SoundStudio and set it to use Toccata, all works fine.
Toccata soundcard shows up in ShowConfig.
This is all on an AmigaOS 3.2.2.1 installation.

**Comment by chris-y** [comment]
It might be worth adding a toggle to the Audio & Video section of the config to enable/disable the soundcard?

**Comment by robinsonb5** [comment]
Thanks for the feedback - glad to hear it's working OK for people.

A toggle might be a good idea - but I'd be very surprised if the card's presence caused problems even with very old and badly written software - it should just sit there at 0xe90000 and be completely inert until software starts poking at it.

**Comment by limi** [comment]
Yeah, I don’t think a toggle is necessary, since it’s not doing anything unless software explicitly asks for it.

I could *maybe* see the argument for having the toggle as a way to say “yes, this core does support 16-bit AHI sound”, but that doesn’t seem like enough reason to add another setting.

Great work, thanks for doing this!

**Comment by apolkosnik** [comment]
I would try to test it in the M68000 mode, since most people run ot on the 020. Other than that I agree that it shouldn't affect the functionality. I don't think there's an explicit switch for enabling/disabling RTG either.

**Comment by limi** [comment]
I don’t think enabling AHI support when running 68000 is really relevant (even though it it technically possible) — as the documentation states:

> Very limited volime control when using 16 bit samples
> No HiFi mixing routines are available.
> No stereo samples
> No DSP effects are available.
> Less precision on some calculations
> Slower
> 
> As you can see, this version is very primitive.  Get yourself an
> accelerator!  I cannot guarantee that I will continue to support this
> processor in the future.

I don’t think making AHI work with 68000 selected in the core should be something we care about, nor should it block anything. And, since it doesn’t touch any existing parts of the architecture — it needs to be explicitly addressed from software — I don’t see how it could interfere with it either. But I’ll let @robinsonb5 make that call, since I only have a high-level understanding of how it works 🙂

**Comment by chris-y** [comment]
> I would try to test it in the M68000 mode, since most people run ot on the 020. Other than that I agree that it shouldn't affect the functionality. I don't think there's an explicit switch for enabling/disabling RTG either.

Yes, there's no switch for RTG but that doesn't show up as an expansion board either, so it's not the same.  If there's definitely no risk of it interfering then a toggle is unnecessary, I just thought that it might be useful/sensible to be able to virtually unplug it.  I did try booting 1.3 with this core and it seems happy enough.

**Comment by robinsonb5** [comment]
If adding a toggle was as simple as adding something to the core's config string then I wouldn't hesitate - and for most cores that's all it would take.

The Amiga core's a special case, though, and (for historical reasons due to both the core's and the firmware's roots in the original Minimig) most of its menu is embedded in mister-main - so the option would need to be added there, and then communicated to the core. 

So I think it's probably not worth the hassle until or unless we find a piece of software which breaks due to the presence of the card.

**Comment by limi** [comment]
Great! Is this ready to merge then, or are we waiting for Paul to chime in on the FIFO fix?

**Comment by robinsonb5** [comment]
Yeah, I'm hoping @ranzbak will merge retrofun's mono fix and also my FIFO fix (or fix the same issue a different way if he prefers) - so that the submodule can reference his repo directly instead of my fork of it, so that the credit's obviously his.

Apart from that, I'd be curious to know if the issues I see with Toccata and RTG on MiST affect MiSTer too.  (I don't have an RTG setup on MiSTer yet - but if I set a module playing in OctaMED SoundStudio and then do something RTG-intensive like flipping screens or opening or moving windows the music stops.  Do you see the same thing on MiSTer?  I think it's just that the RTG system holds up the player for too long, so the player detects an audio buffer underrun and stops, and if so there's probably not much I can do about it.)

So apart from that, I think it's ready?

**Comment by limi** [comment]
FIFO fix was merged: https://github.com/ranzbak/fpga-toccata/pull/2

Toccata and MiSTer RTG seems to be fine, I can flip between screens without any interruption — the only time it happens is when I switch between different resolutions, but I think that’s my TV re-syncing to the new resolution, and probably drops the HDMI audio while it’s doing that. Normal flipping of screens and moving of windows works fine, no interruption to the AHI output.

**Comment by ranzbak** [comment]
Sorry for the delay, I indeed merged both of the pull requests.

**Comment by robinsonb5** [comment]
Thanks @limi (and @ranzbak for merging - no need to apologise - we're doing this for fun, after all!)

I've updated my own repo to point to the up-to-date upstream so I can open a PR now.

**Comment by limi** [comment]
I assume this issue can be closed as fixed?

**Comment by robinsonb5** [comment]
I'd assume so - I don't think there are any outstanding problems with it?
